CoffeaAnalysis/HNLAnalysis/helpers.py

```diff
@@ -0,0 +1,575 @@
+    Sel_Tau = reco_tau[ak.argmin(delta_r(trigger_tau, reco_tau), axis=-1, keepdims = True)]
+    Sel_Trigger_Tau = trigger_tau[ak.argmin(delta_r(trigger_tau, reco_tau), axis=-1, keepdims = True)] 
+    #then we take min dr between reco and trigger overall the best matching trigger tau
+    Sel_Trigger_Tau2 = Sel_Trigger_Tau[ak.argmin(delta_r(Sel_Tau, Sel_Trigger_Tau), axis=1, keepdims = False)][:,0]
+    Sel_Tau2 = Sel_Tau[ak.argmin(delta_r(Sel_Tau, Sel_Trigger_Tau), axis=1, keepdims = False)][:,0]
+
+    #here we can require a min dr between trigger and reco
+
+    # impose min dr between l1 and l2
+    cut_dr_mask = (delta_r(Sel_Tau1, Sel_Tau2) >= min_dr_cut)[:,0]
+    Sel_Tau1 = Sel_Tau1[cut_dr_mask]
+    Sel_Tau2 = Sel_Tau2[cut_dr_mask]
+    events = events[cut_dr_mask]
+
+    # impose min pt > 40 GeV for Tau1 and Tau 2
+    cut_pt_mask = ak.any((Sel_Tau1.pt > min_pt_cut) & (Sel_Tau2.pt > min_pt_cut), axis=1)
+    Sel_Tau1 = Sel_Tau1[cut_pt_mask]
+    Sel_Tau2 = Sel_Tau2[cut_pt_mask]
+    events = events[cut_pt_mask]
+    
+    return events, Sel_Tau1, Sel_Tau2
+
+def select_lep2(events, Sel_Lep1, type, delta_r_cut = 0.5):
+
+    if type not in ['tau','muon', 'electron']:
+        raise 'Unvalid type: must be: tau,muon or electron'
+
+    if type == 'tau':
+        Reco_lep2 = events.SelTau
+
+    if type == 'muon':
+        Reco_lep2 = events.SelMuon
+
+    if type == 'electron':
+        Reco_lep2 = events.SelElectron
+
+    # select reco lep2 with dr(lep1,lep2)>0.5 
+    sel_lep1, reco_lep2= ak.unzip(ak.cartesian([Sel_Lep1, Reco_lep2], nested=True))
+    Sel_Lep2 = reco_lep2[(delta_r(sel_lep1,reco_lep2)> delta_r_cut)]
+
+    #We need only one tau: in case there is more, we choose the one with higher pt 
+    Sel_Lep2 = Sel_Lep2[ak.max(Sel_Lep2.pt, axis=-1) == Sel_Lep2.pt]
+    Sel_Lep2 = ak.fill_none(Sel_Lep2[:,0], [], axis=0)
+
+    return Sel_Lep2
+
+def select_lep3(events, Sel_Lep1, Sel_Lep2, type, delta_r_cut = 0.5):
+
+    if type not in ['tau','muon', 'electron']:
+        raise 'Unvalid type: must be: tau,muon or electron'
+
+    if type == 'tau':
+        Reco_lep3 = events.SelTau
+
+    if type == 'muon':
+        Reco_lep3 = events.SelMuon
+
+    if type == 'electron':
+        Reco_lep3 = events.SelElectron
+
+    # select reco lep3 that is not the same as the lep1 and lep2
+    sel_lep1, reco_lep3 = ak.unzip(ak.cartesian([Sel_Lep1, Reco_lep3], nested=True))
+    match1 = (delta_r(sel_lep1,reco_lep3)> delta_r_cut)
+
+    sel_lep2, reco_lep3 = ak.unzip(ak.cartesian([Sel_Lep2, Reco_lep3], nested=True))
+    match2 = (delta_r(sel_lep2,reco_lep3)> delta_r_cut)
+    Sel_Lep3 = reco_lep3[match1 & match2]
+
+    #We need only one l3: in case there is more, we choose the one with higher pt 
+    Sel_Lep3 = Sel_Lep3[ak.max(Sel_Lep3.pt, axis=-1) == Sel_Lep3.pt]
+    Sel_Lep3 = ak.fill_none(Sel_Lep3[:,0], [], axis=0)
+
+    return Sel_Lep3
+
+def bjet_veto(events, Lepton1, Lepton2, Lepton3, delta_r_cut = 0.5):
```

Comment:
```suggestion
def bjet_veto(events, Lepton1, Lepton2, Lepton3, btagDeepFlavB_cut, delta_r_cut = 0.5):
```
cut on deepFlavB is era dependent